Ethereum 2.0 Networking Specification (#1328)
ethereum/eth2.0-specs: specs/networking/p2p-interface.md

# Overview

This document contains the network specification for Ethereum 2.0 clients.

It consists of four main sections:

1. A specification of the network fundamentals detailing the two network configurations: interoperability test network, and mainnet launch.

> 2. A specification of the three network interaction _domains_ of ETH2.0: (a) the gossip domain, (b) the discovery domain, \(c\) the Req/Resp domain.

Reviewer: Instead of Req/Resp domain, maybe change it to RPC domain?

- `RQRP_MAX_SIZE`: The max size of uncompressed req/resp messages that clients will allow.
 Value: TBD
- `GOSSIP_MAX_SIZE`: The max size of uncompressed gossip messages
 Value: 1MB (estimated from expected largest uncompressed block size).
- `SHARD_SUBNET_COUNT`: The number of shard subnets used in the gossipsub protocol.
 Value: TBD

## The gossip domain: gossipsub

Clients MUST support the [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) libp2p protocol.

**Protocol ID:** `/meshsub/1.0.0`

**Gossipsub Parameters**

*Note: Parameters listed here are subject to a large-scale network feasibility study.*

The following gossipsub parameters will be used:

- `D` (topic stable mesh target count): 6
- `D_low` (topic stable mesh low watermark): 4
- `D_high` (topic stable mesh high watermark): 12
- `D_lazy` (gossip target): 6
- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 60
- `gossip_advertise` (number of windows to gossip about): 3
- `gossip_history` (number of heartbeat intervals to retain message IDs): 5
- `heartbeat_interval` (frequency of heartbeat, seconds): 1

### Topics

> Topics are plain UTF-8 strings, and are encoded on the wire as determined by protobuf (gossipsub messages are enveloped in protobuf messages).

Reviewer: Is there a rationale for storing topics as plain UTF-8 strings instead of topic hashes?

Reviewer 2: Good point. Yes. I will add the response to the rationale section and crosslink the commit here.

Reviewer 2: Actually, it was already there. But I took the opportunity to enhance the answer anyway: https://github.com/status-im/eth2.0-specs/commit/1ba03b4c33a6ade51dbd60dae9a6a15acd7be531

Reviewer 2: For keeping the comment reel tidy, this thread can be marked as resolved @arnetheduck (I can't do it).

Author: breathe in, breathe out :)

| shard{N}\_beacon_attestation | Attestation       |
| voluntary_exit               | VoluntaryExit     |
| proposer_slashing            | ProposerSlashing  |
| attester_slashing            | AttesterSlashing  |

Clients MUST reject (fail validation) messages containing an incorrect type, or invalid payload.

When processing incoming gossip, clients MAY descore or disconnect peers who fail to observe these constraints.

### Encodings

Topics are post-fixed with an encoding. Encodings define how the payload of a gossipsub message is encoded.

#### Interop

- `ssz` - All objects are SSZ-encoded. Example: The beacon block topic string is: `/beacon_block/ssz` and the data field of a gossipsub message is an ssz-encoded `BeaconBlock`.

#### Mainnet

- `ssz_snappy` - All objects are ssz-encoded and then compressed with snappy. Example: The beacon attestation topic string is: `/beacon_attestation/ssz_snappy` and the data field of a gossipsub message is an `Attestation` that has been ssz-encoded then compressed with snappy.

Implementations MUST use a single encoding. Changing an encoding will require coordination between participating implementations.

> ## The discovery domain: discv5

Reviewer: I know this has been discussed in the past but just to try to get a definitive answer: the current discv5 spec uses RLP and the current eth2 networking spec depends on SSZ. Should we simply choose 1 serialization format for all networking or keep both?

Reviewer 2: Alternatively, the ENR spec can be extended to support multiple serialisation formats, e.g. by pretending a prefix that identifies the encoding.

Author: multiple serialization formats means multiple signatures, and having everyone that publishes enr records understand all formats anyway, if I'm not mistaken. no real win. 

what perhaps would make more sense would be to use protobuf or some other established extensible format for enr's given that the requirements are significantly different from blockchain data, but that's probably unpalatable for eth1. regardless, this seems off topic for this thread and belongs in the discv5 forum, as rlp has been deemed unworkable for eth2.

Reviewer 2: Opened a discussion: https://github.com/ethereum/devp2p/issues/107, and added to the open points tracker comment here: https://github.com/ethereum/eth2.0-specs/pull/1328#issuecomment-517664569.

Discovery Version 5 ([discv5](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md)) is used for peer discovery, both in the interoperability testnet and mainnet.

> `discv5` is a standalone protocol, running on UDP on a dedicated port, meant for peer discovery only. `discv5` supports self-certified, flexible peer records (ENRs) and topic-based advertisement, both of which are (or will be) requirements in this context.

Reviewer: We should define what this dedicated port is ahead of time. Maybe have different ports for interop and mainnet

Reviewer 2: Not necessary to define fixed ports. You only need to know the details of the bootstrappers. Past that, nodes announce their UDP ports in their self-certified ENR.

Reviewer: Ahh right! ENRs have a field for both TCP and UDP ports.

# Overview

This document contains the network specification for Ethereum 2.0 clients.

It consists of four main sections:

1. A specification of the network fundamentals detailing the two network configurations: interoperability test network, and mainnet launch.
2. A specification of the three network interaction _domains_ of ETH2.0: (a) the gossip domain, (b) the discovery domain, \(c\) the Req/Resp domain.

> 3. The rationale and further explanation for the design choices made in the previous two sections.

Author: as this lists sections, it reads just fine without the `are` I think

- [Design Decision Rationale](#design-decision-rationale)
  - [Transport](#transport-1)
  - [Multiplexing](#multiplexing-1)
  - [Protocol Negotiation](#protocol-negotiation-1)
  - [Encryption](#encryption)
  - [Gossipsub](#gossipsub)
  - [Req/Resp](#reqresp)
  - [Discovery](#discovery)
  - [Compression/Encoding](#compressionencoding)
- [libp2p Implementations Matrix](#libp2p-implementations-matrix)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Network Fundamentals

This section outlines the specification for the networking stack in Ethereum 2.0 clients.

Sections that have differing parameters for mainnet launch and interoperability testing are split into subsections. Sections that are not split have the same parameters for interoperability testing as mainnet launch.

## Transport

Even though libp2p is a multi-transport stack (designed to listen on multiple simultaneous transports and endpoints transparently), we hereby define a profile for basic interoperability.

#### Interop

All implementations MUST support the TCP libp2p transport, and it MUST be enabled for both dialing and listening (i.e. outbound and inbound connections).

> The libp2p TCP transport supports listening on IPv4 and IPv6 addresses (and on multiple simultaneously). Clients SHOULD allow the operator to configure the listen IP addresses and ports, including the addressing schemes (IPv4, IPv6).

Reviewer 3: Why `SHOULD`? the ability to configure listen IP/ports sounds like a `MUST` to me

Author: actually, I think we can simply remove that sentence. what configuration clients offer is domain-specific to that client.

To facilitate connectivity, and avert possible IPv6 routability/support issues, clients participating in the interoperability testnet MUST expose at least ONE IPv4 endpoint.

> All listening endpoints must be publicly dialable, and thus not rely on libp2p circuit relay, AutoNAT or AutoRelay facilities.

Reviewer 4: Why not? Is there some major blocker to dialing a node via circuit relay? 

We utilize circuit relay for our testnet orchestration, but we can hardcode inbound routes to meet this requirement if needed.

Reviewer 5: IMO, relays add complexity at this stage, but should be eventually incorporated as they play an important role in standing up well interconnected network. 

Here are some questions I'd ask:

- how are relays discovered?
- what are the security implications?
- are regular peers supposed to provide this functionality or is it backed by standalone infra, or both?

Author: for testing purposes, and for those that don't have `auto*` features implemented, it's easier to require a straightforward networking setup to start with - these exotic extras are essentially workarounds for when networking is difficult - not a complication we necessarily want to have from the start, when testing things.

Reviewer 2: Agree with the above. Let's add network topology/connectivity improvements/complexity incrementally.

Reviewer 2: By phasing these aspects in, the barrier for WIP libp2p implementations to become _feature complete_ also becomes more lower and more attainable.

Reviewer 4: Sounds good to me. We'll want relay support soon, but can work without it for interop testing.

Nodes operating behind a NAT, or otherwise undialable by default (e.g. container runtime, firewall, etc.), MUST have their infrastructure configured to enable inbound traffic on the announced public listening endpoint.

#### Mainnet

All requirements from the interoperability testnet apply, except for the IPv4 addressing scheme requirement.

At this stage, clients are licensed to drop IPv4 support if they wish to do so, cognizant of the potential disadvantages in terms of Internet-wide routability/support. Clients MAY choose to listen only on IPv6, but MUST retain capability to dial both IPv4 and IPv6 addresses.

Usage of circuit relay, AutoNAT or AutoRelay will be specifically re-examined closer to the time.

## Encryption and identification

#### Interop

[SecIO](https://github.com/libp2p/specs/tree/master/secio) with `secp256k1` identities will be used for initial interoperability testing.

The following SecIO parameters MUST be supported by all stacks:

-  Key agreement: ECDH-P256.
-  Cipher: AES-128.

> -  Digest: SHA256.

Reviewer 6: nitpick: using name `SHA-256` in this spec.
```suggestion
-  Digest: SHA-256.
```

#### Mainnet

[Noise Framework](http://www.noiseprotocol.org/) handshakes will be used for mainnet. libp2p Noise support [is in the process of being standardised](https://github.com/libp2p/specs/issues/195) in the libp2p project.

> Noise support will presumably include IX, IK and XX handshake patterns, and may rely on Curve25519 keys, ChaCha20 and Poly1305 ciphers, and SHA256 as a hash function. These aspects are being actively debated in the referenced issue [ETH 2.0 implementers are welcome to comment and contribute to the discussion.]

Reviewer 6: ditto
```suggestion
Noise support will presumably include IX, IK and XX handshake patterns, and may rely on Curve25519 keys, ChaCha20 and Poly1305 ciphers, and SHA-256 as a hash function. These aspects are being actively debated in the referenced issue [ETH 2.0 implementers are welcome to comment and contribute to the discussion.]
```

## Multiplexing

During connection bootstrapping, libp2p dynamically negotiates a mutually supported multiplexing method to conduct parallel conversations. This applies to transports that are natively incapable of multiplexing (e.g. TCP, WebSockets, WebRTC), and is omitted for capable transports (e.g. QUIC).

Two multiplexers are commonplace in libp2p implementations: [mplex](https://github.com/libp2p/specs/tree/master/mplex) and [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). Their protocol IDs are, respectively: `/mplex/6.7.0` and `/yamux/1.0.0`.

Clients MUST support [mplex](https://github.com/libp2p/specs/tree/master/mplex) and MAY support [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). If both are supported by the client, yamux must take precedence during negotiation. See the Rationale section of this document for tradeoffs.

# ETH2 network interaction domains

> ## Constants

Author: 514ff5814f233836eba8b45d9bb5d94b97d7a60e

> This section outlines constants that are used in this spec.

Reviewer 4: Please add timeout constants in this section as well. TTFB timeout and response timeout.

Author: 514ff5814f233836eba8b45d9bb5d94b97d7a60e

> - `RQRP_MAX_SIZE`: The max size of uncompressed req/resp messages that clients will allow.

Reviewer 4: nit: what is RQRP? Can this be MSG_MAX_SIZE?

Reviewer 7: looks like an abbreviation for ReQuestResPonse

Reviewer 2: Guilty as charged.`MSG` is ambiguous though; other layers also deal with messages.

Author: I'm not too happy with Req/Resp in general, it's quite the mouthful, happy to hear alternatives.. `DM_MAX_SIZE` is one for direct messaging between peers, as opposed to gossip/broadcast.

Reviewer 2: Naming things is hard 😛 How about we think outside the box of messaging patterns, and describe this domain functionally:

1. "the direct domain" / `DIRECT_MAX_SIZE`
2. "the service domain" / `SVC_MAX_SIZE`
3. "the operation domain" / `OP_MAX_SIZE`

Personally, I like _the service domain_, as it accurately describes what functionality this offers: peers offer services to one another, where one of those services is the synchronisation facility.

WDYT?

Author: Did another round of thinking here, and I'm leaning towards `REQ_MAX_SIZE` for the req/resp domain - services sounds like web services or the like which are quite different in style from this - in particular web services tend to be unidirectional.

We have the pub/sub duality and I suspect the req/resp is another example of the same, leaning on the terminology of p2p message buses like 0mq.

Reviewer 2: `REQ_` => my understanding was that this limitation applied to responses too.

Author: yeah, but it's the request domain which implies a response. not perfect, agree.

Reviewer 7: so whats wrong w/ `REQ_RESP_MAX_SIZE`? I think the stumbling block is less a perfect name and more that `RQRP` is not a commonly used abbreviation (as far as i have seen)

Reviewer 2: > in particular web services tend to be unidirectional

That isn't true. Both REST and SOAP are inherently bidirectional; with or without payloads, they're synchronous.

**Service domain** describes _what_ this domain does, **Req/resp domain** describes _how_ it does it in terms of messaging patterns.

We name other domains with the _what_, not with the _how_, so applying that consistency here would be fair.

Anyway, I'm fine whichever way; I do confess I never really liked req/resp to be honest 😛

Author: at this point, I guess it's less about choosing a good name and more about choosing one that is not too bad :) `REQ_RESP_MAX_SIZE` gets bonus points for staying true to eth2's penchant for excessively long names.

I hold commit rights and take bribes.

 Value: TBD
- `GOSSIP_MAX_SIZE`: The max size of uncompressed gossip messages
 Value: 1MB (estimated from expected largest uncompressed block size).
- `SHARD_SUBNET_COUNT`: The number of shard subnets used in the gossipsub protocol.
 Value: TBD

## The gossip domain: gossipsub

Clients MUST support the [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) libp2p protocol.

**Protocol ID:** `/meshsub/1.0.0`

**Gossipsub Parameters**

*Note: Parameters listed here are subject to a large-scale network feasibility study.*

The following gossipsub parameters will be used:

> - `D` (topic stable mesh target count): 6

Reviewer 4: Why is this called `D`? Other parameters have well formed names, can these have full names as well?

Reviewer: I think `D` is the degree of the gossipsub overlay network.

Reviewer 8: It was to be consistent with the specification. Didn't want to name it something different to the spec as it may add confusion. See here: https://github.com/libp2p/specs/tree/master/pubsub/gossipsub#meshsub-an-overlay-mesh-router
And here: https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L22

Reviewer 2: Yeah. That said, I think the gossipsub spec could do a better job at variable naming. Especially now that it's being cross-referenced from the interwebs. I'll file an issue there.

Author: I'm sticking a link to that spec in there in the meantime

- `D_low` (topic stable mesh low watermark): 4
- `D_high` (topic stable mesh high watermark): 12
- `D_lazy` (gossip target): 6
- `fanout_ttl` (ttl for fanout maps for topics we are not subscribed to but have published to, seconds): 60
- `gossip_advertise` (number of windows to gossip about): 3
- `gossip_history` (number of heartbeat intervals to retain message IDs): 5
- `heartbeat_interval` (frequency of heartbeat, seconds): 1

### Topics

Topics are plain UTF-8 strings, and are encoded on the wire as determined by protobuf (gossipsub messages are enveloped in protobuf messages).

Topic strings have form: `/eth2/TopicName/TopicEncoding`. This defines both the type of data being sent on the topic and how the data field of the message is encoded. (Further details can be found in [Messages](#Messages)).

There are two main topics used to propagate attestations and beacon blocks to all nodes on the network. Their `TopicName`'s are:

> - `beacon_block` - This topic is used solely for propagating new beacon blocks to all nodes on the networks. Blocks are sent in their entirety. Clients who receive a block on this topic MUST validate the block proposer signature before forwarding it across the network.

Reviewer 3: we probably should also validate the block is not older than certain slot (i.e. current finalized slot)

Reviewer 2: Yes, adding such a validation rule reduces DoS vectors.

Author: that is already implied - once finalization, all blocks not on the canonical chain before it become invalid by definition.

> - `beacon_attestation` - This topic is used to propagate aggregated attestations (in their entirety) to subscribing nodes (typically block proposers) to be included in future blocks. Similarly to beacon blocks, clients will be expected to perform some sort of validation before forwarding, but the precise mechanism is still TBD.

Reviewer 3: what validation to be specific? signature? slot?

Reviewer 9: Im thinking a signature validation but only require nodes to randomly perform to reduce overhead but still provide heard immunity from bad attestations. Need to consider this more before making a concrete proposal

Author: a reasonable start is not to propagate orphans (missing parent block) - once you're there, it seems pretty reasonable to require sig validation also. the concern is that constructing the state for the attestation is too costly?

Reviewer 9: > a reasonable start is not to propagate orphans (missing parent block)

true. worth adding this IMO with a note on further validation.

> the concern is that constructing the state for the attestation is too costly

that plus the signature validation on all of the unaggregated (and aggregated) attestations might reduce our ability to gossip as quickly as necessary. Thinking the random validation with heard immunity is the way to go but need to do some modeling.

Author: a problem with random validations is that you can no longer disconnect peers that send you junk / police that they're doing their job

Author: https://github.com/ethereum/eth2.0-specs/issues/1332

Additional topics are used to propagate lower frequency validator messages. Their `TopicName`’s are:

- `voluntary_exit` - This topic is used solely for propagating voluntary validator exits to proposers on the network. Voluntary exits are sent in their entirety. Clients who receive a voluntary exit on this topic MUST validate the conditions within `process_voluntary_exit` before forwarding it across the network.
- `proposer_slashing` - This topic is used solely for propagating proposer slashings to proposers on the network. Proposer slashings are sent in their entirety. Clients who receive a proposer slashing on this topic MUST validate the conditions within `process_proposer_slashing` before forwarding it across the network.
- `attester_slashing` - This topic is used solely for propagating attester slashings to proposers on the network. Attester slashings are sent in their entirety. Clients who receive an attester slashing on this topic MUST validate the conditions within `process_attester_slashing` before forwarding it across the network.

#### Interop

> Unaggregated attestations from all shards are sent to the `beacon_attestation` topic.

Reviewer 4: What about aggregated attestations? They also go to `beacon_attestation` topic in interop config?

Reviewer 8: For lighthouse, during initial interop, we were planning on not doing aggregation (just to simplify initial compatibility). As interop progresses I think yes, we should be sending aggregated attestations to `beacon_attestation`.

Author: Indeed, the aggregation strategy has yet to be decided. An open question is whether to allow aggregated attestations here explicitly - it's fine for nimbus, but would require that all testnet participants can at least deal with reading aggregated attestations. Shout if this is a problem, or we'll allow aggregated also.

Reviewer 9: >  it's fine for nimbus, but would require that all testnet participants can at least deal with reading aggregated attestations

these are the exact some datastructure so I think readability should be fine. I lean toward being able to broadcast aggregates on this same channel for interop

> the aggregation strategy has yet to be decided.

I need to get the "naive" strategy in the validator (or network) spec asap. (subset of committee aggregates and broadcasts to `beacon_`  channel)

Author: > I need to get the "naive" strategy in the validator (or network) spec asap.

It's not a blocker for interop, but indeed a high-priority concern so we can start investigating performance. 

https://github.com/ethereum/eth2.0-specs/issues/1331

The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:

| Topic                        | Message Type      |
|------------------------------|-------------------|
| beacon_block                 | BeaconBlock       |
| beacon_attestation           | Attestation       |
| shard{N}\_beacon_attestation | Attestation       |
| voluntary_exit               | VoluntaryExit     |
| proposer_slashing            | ProposerSlashing  |
| attester_slashing            | AttesterSlashing  |

Clients MUST reject (fail validation) messages containing an incorrect type, or invalid payload.

When processing incoming gossip, clients MAY descore or disconnect peers who fail to observe these constraints.

### Encodings

Topics are post-fixed with an encoding. Encodings define how the payload of a gossipsub message is encoded.

#### Interop

- `ssz` - All objects are SSZ-encoded. Example: The beacon block topic string is: `/beacon_block/ssz` and the data field of a gossipsub message is an ssz-encoded `BeaconBlock`.

#### Mainnet

> - `ssz_snappy` - All objects are ssz-encoded and then compressed with snappy. Example: The beacon attestation topic string is: `/beacon_attestation/ssz_snappy` and the data field of a gossipsub message is an `Attestation` that has been ssz-encoded then compressed with snappy.

Reviewer 6: Since topics are plain UTF-8 strings, does it make sense to have a map of `code -> encoding`, so that we can save some space in these topics? e.g., `ssz` is `1` and `ssz_snappy` is `2`, so the topics are `/beacon_attestation/1` and `/beacon_attestation/2` .

Author: if we do this, we have a higher-value target in `beacon_attestation` - ie one could imagine having a fourcc here where the first three chars are topic and the last is encoding or something like this. that said, if we get stream compression, that should take care of it.

Also, we're sending SSZ - it's not the most bandwidth-friendly format out there already.. not sure this will make a practical difference.

all that said, if there's strong support, I'm fine with indices too.

> Implementations MUST use a single encoding. Changing an encoding will require coordination between participating implementations.

Reviewer 4: Can't we handle encoding support in protocol negotiation? Seems like it would be a good idea to require clients to support ssz snappy, but allow them to support other encoding if they choose.

Reviewer 8: The rationale for this was that segregating topics is similar to segregating nodes based on protocol id. Currently, nodes should not support two encodings. If they did, they would form two segregated networks and blocks propagating from one network to the other would be dependent on relay nodes (nodes that support both) and this can cause all kinds of headaches. 
For now, I think its best to avoid having relay/translation nodes and disjoint networks and just enforce everyone agrees on one compression. If the compression changes, we coordinate a hard fork.

Reviewer 2: Yeah. Multiple encodings cause fragmentation, split brain, chimera; essentially diverging states. It may be a desirable property, but implementing it this way opens a can of worms, which can be addressed, but the cost/benefit isn’t clear to me at this point.

Author: We can still allow clients to publish multiple encodings as long as they at least publish the mandatory one - need a phrasing for that.

Reviewer 10: Pubsub topics should be related to a type of content that is disseminating. In our case message encoding is not a part of the content itself, it rather a transport specific thing that matters between a pair of peers speaking to each other. Therefore, specifying encoding in topic name looks odd and puts additional complexity to the process of its update. It would require either multiple topic distribution from at least _some_ peers in the network or a hard fork coordination which could be quite a tricky thing. We might not want to have a deal with one or another.

Is there a rationale behind _not_ including encoding into a protocol?

Author: it's in the topic name because we're describing about the payload. 

broadcasts are generally not negotiated dynamically between two peers because that would require relaying/translation between different encodings and breaking gossip message identity (which determines if a message is retransmitted to neighbours or not). instead we're opting for a network-wide mandated encoding which will be upgraded only in a coordinated way (for example together with a hard fork). _insert radio station analogy_

there's a separate discussion to solve compression at the transport layer, but for the purposes of this spec, we're not there yet.

Reviewer 10: > and breaking gossip message identity

Looks like I'm starting to get into this. We would have to have a deal with multiple identities for the same message in order to be able to gossip them across peers with different encodings. That is the point, right?

Author: well, different encodings on different topics means the message would have to be re-encoded by relays to reach the peers that listen only on the "other" encoding - this is the fragmentation that @raulk mentions above. in the process, you get differences in identity, so to know if a message has been published on the "other" encoding you have to reencode it first and check if you've seen it already. 

this is quite costly and non-trivial, so instead we're opting for mandating that all gossip happens on one topic with one encoding and upgrades are done "network-wide". 

also, gossip protocols typically remember what messages they've seen for a finite period of time - if you publish the same message again after that time has passed, it will be re-broadcast - adding a relay delay makes this scenario more likely (the identity problem, from a different angle).

one can imagine that in some complicated upgrade scenario, we might have peers publishing the same message on two topics/encodings, but the price here is pretty high in terms of overhead - both computational and networking, so we'd rather avoid that.

Reviewer 10: > this is quite costly and non-trivial, so instead we're opting for mandating that all gossip happens on one topic with one encoding and upgrades are done "network-wide".

Would you mind to add a couple of sentences with this detail to the rationale section relating to topic encodings to prevent arising such questions from future readers?

Author: 514ff5814f233836eba8b45d9bb5d94b97d7a60e

#### Responding side

Once a new stream with the protocol ID for the request type has been negotiated, the responder must process the incoming request message according to the encoding strategy, until EOF (denoting stream half-closure by the requester).

The responder MUST:

1. Use the encoding strategy to read the optional header.
2. If there are any length assertions for length `N`, it should read exactly `N` bytes from the stream, at which point an EOF should arise (no more bytes). Should this is not the case, it should be treated as a failure.
3. Deserialize the expected type, and process the request.
4. Write the response (result, optional header, payload).
5. Close their write side of the stream. At this point, the stream will be fully closed.

If steps (1), (2) or (3) fail due to invalid, malformed or inconsistent data, the responder MUST respond in error. Clients tracking peer reputation MAY record such failures, as well as unexpected events, e.g. early stream resets.

The entire request should be read in no more than **5 seconds**. Upon a timeout, the responder SHOULD reset the stream.

> The responder SHOULD send a response promptly, starting with a **single-byte** response code which determines the contents of the response (`result` particle in the BNF grammar above).

Reviewer 4: We can assume this byte represents an unsigned integer, but can you clarify it explicitly?

Reviewer 2: Yes.

Author: that's what a `byte` generally is though, as opposed to an int8 for example - an octet of bits with no sign really. how should it be clarified? 

https://en.wikipedia.org/wiki/Byte:

> The modern de facto standard of eight bits, as documented in ISO/IEC 2382-1:1993, is a convenient power of two permitting the binary-encoded values 0 through 255 for one byte (2 to the power 8 is 256).

Reviewer 4: it's a small addition to increase clarity, maybe it is intuitive to most people but it wasn't clear to me.

Author: 514ff5814f233836eba8b45d9bb5d94b97d7a60e

It can have one of the following values:

> -  0: **Success** -- a normal response follows, with contents matching the expected message schema and encoding specified in the request.

Reviewer 4: Zero is a default value in some languages, can we set 0 to be UNKNOWN or forbid 0?

Author: 0 = "no error" - seems appropriate, given that the other codes indicate some form of error - else you have to pick some arbitrary value for "success" in a protocol like this - better go with something that's commonly used this way already. Compare protobuf for example, you'd have an "error" enum in there, you'd set 0 as success because that's what the generated code gives you when the error field lacks a value.

Reviewer 4: Sounds good to me and I agree with that perspective. Thanks!

1. If `fork_version` doesn’t match the local fork version, since the client’s chain is on another fork. `fork_version` can also be used to segregate testnets.
2. If the (`finalized_root`, `finalized_epoch`) shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 sends (root, epoch) of (A, 5) and Peer 2 sends (B, 3) but Peer 1 has root C at epoch 3, then Peer 1 would disconnect because it knows that their chains are irreparably disjoint.

Once the handshake completes, the client with the lower `finalized_epoch` or `head_slot` (if the clients have equal `finalized_epoch`s) SHOULD request beacon blocks from its counterparty via the `BeaconBlocks` request.

#### Goodbye

**Protocol ID:** ``/eth2/beacon_chain/req/goodbye/1/``

**Content:**
```
(
  reason: uint64
)
```
Client MAY send goodbye messages upon disconnection. The reason field MAY be one of the following values:

- 1: Client shut down.
- 2: Irrelevant network.
- 3: Fault/error.

Clients MAY use reason codes above `128` to indicate alternative, erroneous request-specific responses.

The range `[4, 127]` is RESERVED for future usage.

> #### BeaconBlocks

Reviewer 3: I think we need to add one for `BeaconState` as well, to request the last finalized state to start syncing

Author: for interop, we start with full sync clients from genesis - syncing state should probably be landed together with light clients and a strategy for deciding _which_ states should be available for download (all, certain epochs, one per month, etc)

**Protocol ID:** `/eth2/beacon_chain/req/beacon_blocks/1/`

Request Content
```
(
  head_block_root: HashTreeRoot
  start_slot: uint64
  count: uint64
  step: uint64
)
```

Response Content:
```
(
  blocks: []BeaconBlock
)
```

> Requests count beacon blocks from the peer starting from `start_slot` on the chain defined by `head_block_root`. The response MUST contain no more than count blocks. step defines the slot increment between blocks. For example, requesting blocks starting at `start_slot` 2 with a step value of 2 would return the blocks at [2, 4, 6, …]. In cases where a slot is empty for a given slot number, no block is returned. For example, if slot 4 were empty in the previous example, the returned array would contain [2, 6, …]. A step value of 1 returns all blocks on the range `[start_slot, start_slot + count)`.

Reviewer 4: ```suggestion
Requests count beacon blocks from the peer starting from `start_slot` on the chain defined by `head_block_root`. The response MUST contain no more than count blocks. `step` defines the slot increment between blocks. For example, requesting blocks starting at `start_slot` 2 with a step value of 2 would return the blocks at [2, 4, 6, …]. In cases where a slot is empty for a given slot number, no block is returned. For example, if slot 4 were empty in the previous example, the returned array would contain [2, 6, …]. A step value of 1 returns all blocks on the range `[start_slot, start_slot + count)`.
```

Added back ticks around `step`

`BeaconBlocks` is primarily used to sync historical blocks.

Clients MUST support requesting blocks since the start of the weak subjectivity period and up to the given `head_block_root`.

Clients MUST support `head_block_root` values since the latest finalized epoch.

#### RecentBeaconBlocks

**Protocol ID:** `/eth2/beacon_chain/req/recent_beacon_blocks/1/`

Request Content:

```
(
  block_roots: []HashTreeRoot
)
```

Response Content:

```
(
  blocks: []BeaconBlock
)
```

Requests blocks by their block roots. The response is a list of `BeaconBlock` with the same length as the request. Blocks are returned in order of the request and any missing/unknown blocks are left empty (SSZ null `BeaconBlock`).

`RecentBeaconBlocks` is primarily used to recover recent blocks, for example when receiving a block or attestation whose parent is unknown.

Clients MUST support requesting blocks since the latest finalized epoch.

> 

Reviewer 11: Do we need/want to request recent beacon state(s)? to eg: fetch a latest finalized state for syncing initially

Reviewer 9: State sync is something we certainly should add, but we need to examine a set of tradeoffs along with the current state sync upgrade proposals currently being considered in eth1. 

Fortunately for interop and even for mainnet for a while, we will be fine with a pure block sync. That said, I think we should have something in place or at least a proposal in consideration before mainnet launch

Author: agreed, post-interop at least.

The alternative is to maintain a fully connected mesh (all peers connected to each other 1:1), which scales poorly (O(n^2)).

### Why are we using topics to segregate encodings, yet only support one encoding?

For future extensibility with almost zero overhead now (besides the extra bytes in the topic name).

### How do we upgrade gossip channels (e.g. changes in encoding, compression)?

Such upgrades lead to fragmentation, so they’ll need to be carried out in a coordinated manner most likely during a hard fork.

### Why are the topics strings and not hashes?

Topics names have a hierarchical structure. In the future, gossipsub may support wildcard subscriptions (e.g. subscribe to all children topics under a root prefix) by way of prefix matching. Enforcing hashes for topic names would preclude us from leveraging such features going forward.

No security or privacy guarantees are lost as a result of choosing plaintext topic names, since the domain is finite anyway, and calculating a digest's preimage would be trivial.

> Furthermore, the ETH2 topic names are shorter their digest equivalents (asuming SHA-256 hash), so hashing topics would bloat messages unnecessarily.

Reviewer 2: ```suggestion
Furthermore, the ETH2 topic names are shorter than their digest equivalents (asuming SHA-256 hash), so hashing topics would bloat messages unnecessarily.
```

> Furthermore, the ETH2 topic names are shorter their digest equivalents (assuming SHA-256 hash), so hashing topics would bloat messages unnecessarily.

Reviewer 2: ```suggestion
Furthermore, the ETH2 topic names are shorter than their digest equivalents (assuming SHA-256 hash), so hashing topics would bloat messages unnecessarily.
```

### Why are we sending entire objects in the pubsub and not just hashes?

Entire objects should be sent to get the greatest propagation speeds. If only hashes are sent, then block and attestation propagation is dependent on recursive requests from each peer. In a hash-only scenario, peers could receive hashes without knowing who to download the actual contents from. Sending entire objects ensures that they get propagated through the entire network.

### Should clients gossip blocks if they *cannot* validate the proposer signature due to not yet being synced, not knowing the head block, etc?

The prohibition of unverified-block-gossiping extends to nodes that cannot verify a signature due to not being fully synced to ensure that such (amplified) DOS attacks are not possible.

### How are we going to discover peers in a gossipsub topic?

> Via discv5 topics. ENRs should not be used for this purpose, as they store identity, location and capability info, not volatile advertisements.

Reviewer 10: It would be useful to add a link to Topic Advertisement here.

Reviewer 5: I think it makes sense to have discv5 section here that outlines how eth2.0 will use it specifically.

Author: link added, marked as under construction as discv5 specifics are still being worked out

514ff5814f233836eba8b45d9bb5d94b97d7a60e

In the interoperability testnet, all peers will be subscribed to all global beacon chain topics, so discovering peers in specific shard topics will be unnecessary.

## Req/Resp

> ### Why segregate requests into dedicated protocol IDs?

Reviewer 5: In theory, I like the elegance of `REST` like endpoints, but in practice they might be as ambiguous as `semver`.

For example, how does the client know which protocol the other peer speaks? This isn't easy to do if every endpoint is versioned independently and the client relies on libp2p internals such as `identify` and `multistream/multiselect` to signal which versions the other end supports. Which right now (AFAIR), isn't possible without probing each endpoint explicitly.

Reviewer 2: > For example, how does the client know which protocol the other peer speaks?

We have planned to address that in the discovery layer, but we're deferring defining the exact mechanism because it's irrelevant until mainnet. There's an idea here though: https://github.com/ethereum/eth2.0-specs/pull/1328/files#diff-9e07d955515655364c9a5f28c94f5627R236

Reviewer 2: Opened an issue to track this: https://github.com/ethereum/eth2.0-specs/issues/1335, and added a checklist item to the Open Points comment below: https://github.com/ethereum/eth2.0-specs/pull/1328#issuecomment-517664569.

Requests are segregated by protocol ID to:

> 1. Leverage protocol routing in libp2p, such that the libp2p stack will route the incoming stream to the appropriate handler. This allows each the handler function for each request type to be self-contained. For an analogy, think about how you attach HTTP handlers to a REST API server.

Author: this has been discussed at length in previous pr's, and landed here finally.

there's a half-baked plan to create "bundles" of features for advertisement purposes to manage version complexity, should this become a problem - so that clients can rely on there being a suite of protocols that work well together.

> 2. Version requests independently. In a coarser-grained umbrella protocol, the entire protocol would have to be versioned even if just one field in a single message changed.

Reviewer 5: counterpoint, how to reliably handle interdependent endpoints (are there any?). If endpoint `A` also needs endpoint `B` upgraded, there is no mechanism to do that reliably with this approach.

Reviewer 2: The plan is to group `A` and `B` in the discovery layer under protocol "families/suites/manifests" -- which represent aggregations of finer-grained protocols under one logical unit. that way we get the best of both worlds: upgraded peers can advertise/negotiate newer versions of a given operation, while continuing to serve older peers by supporting the former versions too.

while we could consider a mechanism to expose machine-readable metadata that specifies that `A` depends on `B`, I believe that's beyond scope. in practice, it should be enough to:

1. specify that in the RFC/EIP that introduces `A`.
2. potentially grouping `A` and `B` under a protocol family that can be atomically advertised.
3. allowing peers to incur in a runtime error when they encounter a peer that supports `A` but not `B`.

Reviewer 2: Cross-referenced from https://github.com/ethereum/eth2.0-specs/issues/1335.

> 3. Enable clients to select the individual requests/versions they support. It would no longer be a strict requirement to support all requests, and clients, in principle, could support a subset of equests and variety of versions.

Reviewer 5: typo: `equests` should be `requests`

Reviewer 2: Fixed. @arnetheduck could you mark as resolved?

> 3. Enable clients to select the individual requests/versions they support. It would no longer be a strict requirement to support all requests, and clients, in principle, could support a subset of equests and variety of versions.

Reviewer 2: ```suggestion
3. Enable clients to select the individual requests/versions they support. It would no longer be a strict requirement to support all requests, and clients, in principle, could support a subset of requests and variety of versions.
```

4. Enable flexibility and agility for clients adopting spec changes that impact the request, by signalling to peers exactly which subset of new/old requests they support.
5. Enable clients to explicitly choose backwards compatibility at the request granularity. Without this, clients would be forced to support entire versions of the coarser request protocol.
6. Parallelise RFCs (or ETH2 EIPs). By decoupling requests from one another, each RFC that affects the request protocol can be deployed/tested/debated independently without relying on a synchronisation point to version the general top-level protocol.
  1. This has the benefit that clients can explicitly choose which RFCs to deploy without buying into all other RFCs that may be included in that top-level version.
  2. Affording this level of granularity with a top-level protocol would imply creating as many variants (e.g. /protocol/43-{a,b,c,d,...}) as the cartesian product of RFCs inflight, O(n^2).

> 7. Allow us to simplify the payload of requests. Request-id’s and method-ids no longer need to be sent. The encoding/request type and version can all be handled by the framework.

Reviewer 5: see first item for counterpoint.

### Why are messages length-prefixed with a protobuf varint in the SSZ encoding?

In stream-oriented protocols, we need to delimit messages from one another, so that the reader knows where one message ends and the next one starts. Length-prefixing is an effective solution. Alternatively, one could set a delimiter char/string, but this can readily cause ambiguity if the message itself may contain the delimiter. It also introduces another set of edge cases to model for, thus causing unnecessary complexity, especially if messages are to be compressed (and thus mutated beyond our control).

That said, in our case, streams are single-use. libp2p streams are full-duplex, and each party is responsible for closing their write side (like in TCP). We therefore use stream closure to mark the end of a request.

> Nevertheless, messages are still length-prefixed to prevent DOS attacks where malicious actors send large amounts of data disguised as a request. A length prefix allows clients to set a maximum limit, and once that limit is read, the client can cease reading and disconnect the stream. This allows a client to determine the exact length of the packet being sent, and it capacitates it to reset the stream early if the other party expresses they intend to send too much data.

Author: This is a good point - if somebody is instigating a DoS attack, would a length-prefix hinder them? It seems that they would spoof it anyway.

> Nevertheless, messages are still length-prefixed to prevent DOS attacks where malicious actors send large amounts of data disguised as a request. A length prefix allows clients to set a maximum limit, and once that limit is read, the client can cease reading and disconnect the stream. This allows a client to determine the exact length of the packet being sent, and it capacitates it to reset the stream early if the other party expresses they intend to send too much data.

Reviewer 5: It makes sense when having to manually frame chunks, but in this case we're relying on dedicated streams, so it doesn't seem to add much of anything.

> Nevertheless, messages are still length-prefixed to prevent DOS attacks where malicious actors send large amounts of data disguised as a request. A length prefix allows clients to set a maximum limit, and once that limit is read, the client can cease reading and disconnect the stream. This allows a client to determine the exact length of the packet being sent, and it capacitates it to reset the stream early if the other party expresses they intend to send too much data.

Author: d09d56bec8cec98e3d77a286c48972521dd2be96

### Why do we version protocol strings with ordinals instead of semver?

Using semver for network protocols is confusing. It is never clear what a change in a field, even if backwards compatible on deserialisation, actually implies. Network protocol agreement should be explicit. Imagine two peers:

-  Peer A supporting v1.1.1 of protocol X.
-  Peer B supporting v1.1.2 of protocol X.

These two peers should never speak to each other because the results can be unpredictable. This is an oversimplification: imagine the same problem with a set of 10 possible versions. We now have 10^2 (100) possible outcomes that peers need to model for. The resulting complexity is unwieldy.

For this reason, we rely on negotiation of explicit, verbatim protocols. In the above case, peer B would provide backwards compatibility by supporting and advertising both v1.1.1 and v1.1.2 of the protocol.

Therefore, semver would be relegated to convey expectations at the human level, and it wouldn't do a good job there either, because it's unclear if "backwards-compatibility" and "breaking change" apply only to wire schema level, to behaviour, etc.

> For this reason, we remove semver out of the picture and replace it with ordinals that require explicit agreement and do not mandate a specific policy for changes.

Reviewer 5: It's unclear to me what the versioning policy is - is it strict equality (`==`) or are protocols by default backwards compatible (`<=`)? IMO, this should be explicit in the spec.

Reviewer 8: It's strict equality. 
Support for previous version require the local node to explicitly support a version. You can order which is attempted to be negotiated first, i.e higher version before lower ones.

Reviewer 2: Indeed. I’ll clarify/reword this in the spec.

Author: 514ff5814f233836eba8b45d9bb5d94b97d7a60e

To facilitate connectivity, and avert possible IPv6 routability/support issues, clients participating in the interoperability testnet MUST expose at least ONE IPv4 endpoint.

All listening endpoints must be publicly dialable, and thus not rely on libp2p circuit relay, AutoNAT or AutoRelay facilities.

Nodes operating behind a NAT, or otherwise undialable by default (e.g. container runtime, firewall, etc.), MUST have their infrastructure configured to enable inbound traffic on the announced public listening endpoint.

#### Mainnet

All requirements from the interoperability testnet apply, except for the IPv4 addressing scheme requirement.

At this stage, clients are licensed to drop IPv4 support if they wish to do so, cognizant of the potential disadvantages in terms of Internet-wide routability/support. Clients MAY choose to listen only on IPv6, but MUST retain capability to dial both IPv4 and IPv6 addresses.

Usage of circuit relay, AutoNAT or AutoRelay will be specifically re-examined closer to the time.

## Encryption and identification

#### Interop

> [SecIO](https://github.com/libp2p/specs/tree/master/secio) with `secp256k1` identities will be used for initial interoperability testing.

Reviewer 7: where does `secp256k1` come into play here? if you are referring to the key exchange, i believe the P-256 curve (cf. the Go std lib `crypto/elliptic`) is not the same as the `secp256k1` curve

Reviewer 12: @ralexstokes of course for ECDHE, P-256 curve are used. But for peer identification libp2p supports many algorithms (RSA, ED25519, ECDSA P-256/384/512 and secp256k1). Peer identification used to build `PeerID` and to sign/verify packets in gossipsub protocol.

Reviewer 7: ok thanks -- i thought peer IDs were only RSA, but that is not correct :)

The following SecIO parameters MUST be supported by all stacks:

-  Key agreement: ECDH-P256.
-  Cipher: AES-128.
-  Digest: SHA-256.

#### Mainnet

> [Noise Framework](http://www.noiseprotocol.org/) handshakes will be used for mainnet. libp2p Noise support [is in the process of being standardised](https://github.com/libp2p/specs/issues/195) in the libp2p project.

Reviewer 13: I think you should specify that mainnet clients MUST not connect via secio

Author: We ensure that peers at least support the Noise handshake - naturally clients will gravitate towards it as it will have the most peers available. That said, we want to allow alternate transports and protocols for environments where the default is not feasible.

The expectation is that libp2p clients will gradually be deprioritizing connections through secio naturally as better alternatives take its place (fewer round trips, better security) - there's no need to explicitly forbid it.

> Noise support will presumably include IX, IK and XX handshake patterns, and may rely on Curve25519 keys, ChaCha20 and Poly1305 ciphers, and SHA-256 as a hash function. These aspects are being actively debated in the referenced issue [ETH 2.0 implementers are welcome to comment and contribute to the discussion.]

Reviewer 13: How would this interact with QUIC where the standard defines it's own authentication and confidentiality layer?

Reviewer 13: Also it seems weird to me to support more than one noise handshake pattern.

Reviewer 13: Also if you want to mandate an authenticated handshake like `XX`, how does this interact with `secp256k1` identity key?

Reviewer 13: There is an extension to the noise protocol that supports signatures that may be relevant but to my knowledge none of the noise implementations support it.

https://github.com/noiseprotocol/noise_sig_spec/blob/master/output/noise_sig.pdf

Author: These are excellent points but belong in https://github.com/libp2p/specs/issues/195

Re QUIC, libp2p will understand the features that quic offers and disable its own corresponding protocols accordingly

Reviewer 2: @zmanian this should cover most questions: https://github.com/libp2p/specs/issues/195#issuecomment-517353941

Offering ?K and ?X variants is useful because in some cases you will have the peer’s public static key (e.g. bootstrappers, prev connection). And, as the initiator, you’ll want to leverage it to encrypt your 0-RTT early data.

re: QUIC. Transports offering native encryption/authentication or multiplexing don’t overlay such features.
